_SPIKE_.md

```diff
@@ -0,0 +1,543 @@
+
+- Keep supporting `keys` field as a key-value map from database column name to
+  indexing order, see the description of the actual status below. Entries from
+  `keys` should me merged with entries from `properties`, `keys` taking
+  precedence (replacing `properties` entries).
+
+- Keep supporting `unique` field (set it to `true` to let the index enforce
+  uniqueness).
+
+- Database-specific options will be stored under a key with the connector name:
+
+  ```js
+  {
+    properties: {
+      email: 'ASC',
+    },
+    mongodb: {
+      sparse: true,
+    },
+    mysql: {
+      kind: 'fulltext',
+      type: 'hash',
+    },
+    postgresql: {
+      type: 'hash',
+    }
+  }
+  ```
+
+- To support more complex indexes offered by e.g. PostgreSQL, we should support
+  `columns` field too. This field should contain an array of column-index
+  specification, for example:
+
+  ```js
+  ['email ASC', 'lower(name) DESC'];
+  ```
+
+  In the contrary with the current implementation, I am proposing that `columns`
+  should replace any column list created from `properties` and `keys`, and that
+  `columns` should be nested inside connector-specific options:
+
+  ```js
+  {
+    properties: {
+      // default setup used by most connectors
+      email: 'ASC',
+    },
+    postgresql: {
+      // in PostgreSQL, use a different index definition
+      columns: ['lower(email) ASC']
+    },
+  }
+  ```
+
+### Foreign keys at property level
+
+Introduce a new property metadata "references" (inspired by ANSI SQL):
+
+```js
+{
+  categoryId: {
+    type: 'number',
+    references: {
+      // a TypeResolver
+      model: () => Category,
+      // or a Model class
+      model: Category
+      // or a Model name
+      model: 'Category'
+
+      // name of the target property
+      property: 'id',
+
+      // referential actions
+      onUpdate: 'CASCADE',
+      onDelete: 'CASCADE',
+    }
+  }
+}
+```
+
+Support connector-specific configuration too.
+
+```js
+{
+  categoryId: {
+    type: 'number',
+    mysql: {
+      references: {
+        model: () => Category,
+        property: 'id',
+      }
+    }
+  }
+}
+```
+
+### Foreign keys at model level
+
+Modify the current connector-dependant syntax to make it easier to read and
+support composite foreign keys too.
+
+```js
+{
+  foreignKeys: {
+    [keyName]: {
+      // optional, overrides keyName
+      name: 'constraint_name_for_db',
+
+      // Property name(s) (will be mapped to column name)
+      // formerly: foreignKey
+      sourceProperties: ['source property name'],
+
+
+      // formerly: entity
+      targetModel: 'TargetModel',
+
+      // Property name(s) (will be mapped to column name)
+      // formerly: entityKey
+      targetProperties: ['target property name'],
+
+      // referential actions
+      onUpdate: 'CASCADE',
+      onDelete: 'CASCADE',
+    }
+  }
+}
+```
+
+### Decorators
+
+I am proposing to reuse existing `@property` and `@model` decorators. They are
```

Comment:
+1 on reusing existing decorators from the users' point of view.

Reply:
What about relationship decorators such as `@belongsTo`, `@hasMany`, `@hasOne`. I know we recently added support to add property data to those decorators, but I don't think it is something we want to encourage.

Reply:
See my comment above. The new property-definition field `references` will be usable with the relational decorators supporting custom property definition metadata.

```ts
@model()
class Todo extends Entity {
  // ...

  @belongsTo(
    () => TodoList,
    {},
    {
      /*==== Define a foreign key to enforce referential integrity ===*/
      references: {
        model: () => TodoList,
        property: 'id',
        onDelete: 'CASCADE',
     },
    },
  )
  todoListId: number;
}
```